### PR TITLE
{WIP} docker reorg

### PR DESCRIPTION
**Pull request by robkooper**
Fixes some docker problems:
- fixed issue queues for models

Added new model
- maespa

Cleaner installation of dependencies of PEcAN:
- scripts/generate_makefile_deps.R renamed to scripts/generate_dependencies.R
- This script now generates dependency files for both make as well as the depends image in docker.

Some reorganization of the docker build files
- Moved Dockerfiles from docker /base/* to docker/*/Dockerfile
- thredds now depends on official 4.6 image
- update build files to take into account new docker directory
- script to build/deploy containers this runs on pecan-dev.ncsa.illinois.edu
- some tweaks to install script

## Review Time Estimate

- [ ] Immediately
- [ ] Within one week
- [X] When possible

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue) 
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] My change requires a change to the documentation.
- [X] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by robkooper** [comment]
@infotroph and @ashiklom I think you have most experience with docker. If others have experience with docker I'd love some feedback on the new directory hierarchy.

docker
- depends
-- Dockerfile
-- otherfiles needed
- bae
-- Dockerfile
-- otherfiles needed
...

Only base web docs and need access to full pecan folder to build docker image, and thus will require to copy all files to docker. Others can be run inside the respective folder and will copy a smaller number of files to docker when building, which hopefully speeds up build process.

All models will build with the models/<modelname> folder set as context again to minimize the number of files send to the docker daemon.

Using `install2.r` in Dockerfile, this comes in tidyverse and is command line tool to install packages.

File: scripts/travis/func.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+function travis_time_start {
+    set +x
+    TRAVIS_START_TIME=$(date +%s%N)
+    TRAVIS_TIME_ID=$(cat /dev/urandom | tr -dc 'a-z0-9' | fold -w 8 | head -n 1)
+    TRAVIS_FOLD_NAME=$1
+    echo -e "\e[0Ktraivs_fold:start:$TRAVIS_FOLD_NAME"
+    echo -e "\e[0Ktraivs_time:start:$TRAVIS_TIME_ID"
```
Comment (infotroph) [inline scripts/travis/func.sh]:
```suggestion
    echo -e "\e[0Ktravis_fold:start:$TRAVIS_FOLD_NAME"
    echo -e "\e[0Ktravis_time:start:$TRAVIS_TIME_ID"
```

Comment (infotroph) [reply scripts/travis/func.sh]: ...Bah. Apparently "suggest changes" doesn't know how to handle two-line patches.

File: scripts/travis/func.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    fi
+    set -x
+}
+
+function travis_time_end {
+    set +x
+    _COLOR=${1:-32}
+    TRAVIS_END_TIME=$(date +%s%N)
+    TIME_ELAPSED_SECONDS=$(( ($TRAVIS_END_TIME - $TRAVIS_START_TIME)/1000000000 ))
+    echo -e "traivs_time:end:$TRAVIS_TIME_ID:start=$TRAVIS_START_TIME,finish=$TRAVIS_END_TIME,duration=$(($TRAVIS_END_TIME - $TRAVIS_START_TIME))\n\e[0K"
```
Comment (infotroph) [inline scripts/travis/func.sh]:
```suggestion
    echo -e "travis_time:end:$TRAVIS_TIME_ID:start=$TRAVIS_START_TIME,finish=$TRAVIS_END_TIME,duration=$(($TRAVIS_END_TIME - $TRAVIS_START_TIME))\n\e[0K"
```

File: scripts/travis/func.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+function travis_time_start {
+    set +x
+    TRAVIS_START_TIME=$(date +%s%N)
+    TRAVIS_TIME_ID=$(cat /dev/urandom | tr -dc 'a-z0-9' | fold -w 8 | head -n 1)
+    TRAVIS_FOLD_NAME=$1
+    echo -e "\e[0Ktraivs_fold:start:$TRAVIS_FOLD_NAME"
```
Comment (infotroph) [inline scripts/travis/func.sh]: "traivs" looks like a typo but is consistent through the whole function -- pasted typo or intentional?

File: scripts/travis/compile.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+# TODO: Would probably be faster to use -j2 NCPUS=1 as for other steps
+# but many dependency compilations seem not parallel-safe
+NCPUS=1
+
+for x in base/logger base/remote modules/emulator base/utils base/db base/settings base/visualization base/qaqc modules/data.atmosphere modules/data.land modules/priors modules/uncertainty base/workflow modules/allometry modules/meta.analysis modules/assim.batch modules/assim.sequential modules/benchmark modules/data.hydrology modules/data.remote modules/photosynthesis models/template models/ed modules/rtm models/biocro models/clm45 models/dalec models/dvmdostem models/fates models/gday models/jules models/linkages models/lpjguess models/maat models/maespa models/preles models/sipnet base/all ; do
+    key=$(echo "make/$x" | sed 's#/#_#g')
```
Comment (infotroph) [inline scripts/travis/compile.sh]: This order looks like it's derived from current dependency relationships. Will it need to change when package dependencies change?

**Comment by robkooper** [comment]
Splitting into new pull requests:
#2289 == just docker

**Comment by robkooper** [comment]
@infotroph yeah, I keep messing this up and created a single pull request for docker and will make a new pull request for travis.

I was trying to see if I can add some folds around the build process for each of the sub pieces so I can see where it is slow. The list in travis/compile.sh will be generated automatically.

**Pull request closed without merging**